README.md
[![Build Status][circleci-image]][circleci-url]
[![Go Report Card][reportcard-image]][reportcard-url]
[![Codebeat Status][codebeat-image]][codebeat-url]
[![Coverage][gocover-image]][gocover-url]
[![Total Downloads][gh-downloads-image]][gh-downloads-url]
[![CII Best Practices][cii-bp-image]][cii-bp-url]

# teams

CLI application to manage GitHub issues and pull requests for a team's repositories.

## Installing

1. Get the latest `teams` for your platform from the [releases](https://github.com/hairyhenderson/teams/releases) page
2. Store the downloaded binary somewhere in your path as `teams` (or `teams.exe`
  on Windows)
3. Make sure it's executable (on Linux/macOS)
3. Test it out with `teams --help`!

In other words:

```console
$ curl -o /usr/local/bin/teams https://github.com/hairyhenderson/teams/releases/download/<version>/teams_<os>-<arch>
$ chmod 755 /usr/local/bin/teams
$ teams --help
...
```

_Please report any bugs found in the [issue tracker](https://github.com/hairyhenderson/teams/issues/)._

## Usage

Set the `GITHUB_API_TOKEN` environment variable to a [Personal Access Token](https://github.com/settings/tokens) value. The scopes necessary are `read:org` and `repo`.


## Releasing

Right now the release process is semi-automatic.

1. Create a release tag: `git tag -a v0.0.9 -m "Releasing v0.9.9" && git push --tags`
2. Build binaries & compress most of them: `make build-release`
3. Create a release in [github](https://github.com/hairyhenderson/teams/releases)!

## License

[The MIT License](http://opensource.org/licenses/MIT)

Copyright (c) 2016 Dave Henderson

[circleci-image]: https://img.shields.io/circleci/project/hairyhenderson/teams.svg?style=flat
[circleci-url]: https://circleci.com/gh/hairyhenderson/teams
[reportcard-image]: https://goreportcard.com/badge/github.com/hairyhenderson/teams
[reportcard-url]: https://goreportcard.com/report/github.com/hairyhenderson/teams
[codebeat-image]: https://codebeat.co/badges/55ea20ac-5ff5-4c81-8fd9-8f8c4f18b821
[codebeat-url]: https://codebeat.co/projects/github-com-hairyhenderson-teams
[gocover-image]: https://gocover.io/_badge/github.com/hairyhenderson/teams
[gocover-url]: https://gocover.io/github.com/hairyhenderson/teams
[gh-downloads-image]: https://img.shields.io/github/downloads/hairyhenderson/teams/total.svg
[gh-downloads-url]: https://github.com/hairyhenderson/teams/releases

[cii-bp-image]: https://bestpractices.coreinfrastructure.org/projects/379/badge
[cii-bp-url]: https://bestpractices.coreinfrastructure.org/projects/379

[![Analytics](https://ga-beacon.appspot.com/UA-82637990-1/teams/README.md?pixel)](https://github.com/igrigorik/ga-beacon)


CHANGELOG.md
# Change Log

## [v0.1.1](https://github.com/hairyhenderson/teams/tree/v0.1.1) (2016-09-15)
[Full Changelog](https://github.com/hairyhenderson/teams/compare/v0.1.0...v0.1.1)

**Implemented enhancements:**

- Adding 3 new filters - --user, --state, and --new [\#4](https://github.com/hairyhenderson/teams/pull/4) ([hairyhenderson](https://github.com/hairyhenderson))

**Merged pull requests:**

- Starting to add some tests and extracting some utils [\#3](https://github.com/hairyhenderson/teams/pull/3) ([hairyhenderson](https://github.com/hairyhenderson))
- Adding makefile and hooking up version generation [\#2](https://github.com/hairyhenderson/teams/pull/2) ([hairyhenderson](https://github.com/hairyhenderson))

## [v0.1.0](https://github.com/hairyhenderson/teams/tree/v0.1.0) (2016-09-14)
**Merged pull requests:**

- Initial commit [\#1](https://github.com/hairyhenderson/teams/pull/1) ([hairyhenderson](https://github.com/hairyhenderson))



\* *This Change Log was automatically generated by [github_changelog_generator](https://github.com/skywinder/Github-Changelog-Generator)*

utils_test.go
package main

import (
	"testing"
	"time"

	"github.com/stretchr/testify/assert"
)

func TestHumanDuration(t *testing.T) {
	assert.Equal(t, "Less than a second", HumanDuration(200*time.Millisecond))

	assert.Equal(t, "15 seconds", HumanDuration(15*time.Second))

	assert.Equal(t, "About a minute", HumanDuration(90*time.Second))

	assert.Equal(t, "5 minutes", HumanDuration(5*time.Minute))

	assert.Equal(t, "About an hour", HumanDuration(75*time.Minute))

	assert.Equal(t, "27 hours", HumanDuration(27*time.Hour))

	assert.Equal(t, "3 days", HumanDuration(76*time.Hour))
}


github_test.go
package main

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

func TestRepoString(t *testing.T) {
	r := Repo{
		Org:  "foo",
		Name: "bar",
	}
	assert.Equal(t, "foo/bar", r.String())
}


main.go
package main

import (
	"fmt"
	"log"
	"os"
	"strings"
	"text/tabwriter"

	"github.com/hairyhenderson/teams/version"
	"github.com/jawher/mow.cli"
)

func main() {
	app := cli.App("teams", "Manage GitHub issues and pull requests for a team's repositories")
	app.Version("v version", version.Version)

	teamOpt := app.StringOpt("t team", "", "Team name to manage (in org/teamname format)")

	g := NewGitHubClient()
	var repoList []Repo

	// app.Before =
	app.Before = func() {
		t := strings.SplitN(*teamOpt, "/", 2)
		org := t[0]
		team := t[1]

		repoList, _ = g.GetRepos(org, team)
	}

	app.Command("pulls", "Displays open PRs for the given repo(s)", func(cmd *cli.Cmd) {
		stateArg := cmd.StringOpt("state", "open", "Display PRs based on their state")
		milestoneArg := cmd.StringOpt("m milestone", "", "Milestone to filter on")
		userArg := cmd.StringOpt("u user", "", "Display only PRs from <user>")
		newArg := cmd.BoolOpt("new", false, "Display PRs opened within the past 24 hours")

		cmd.Action = func() {
			w := tabwriter.NewWriter(os.Stdout, 4, 2, 2, ' ', 0)
			fmt.Fprintf(w, "%s\t%s\t%s\t%s\t%s\n", nc("PULL REQUEST"), nc("LAST UPDATED"), nc("CONTRIBUTOR"), nc("MILESTONE"), nc("TITLE"))
			filter := Filter{
				State: *(stateArg),
				New:   *(newArg),
			}
			if milestoneArg != nil {
				filter.Milestone = *(milestoneArg)
			}
			if userArg != nil {
				filter.User = *(userArg)
			}
			for _, repo := range repoList {
				pulls, _ := g.GetPRs(repo, filter)
				g.DisplayPullRequests(w, repo, pulls)
			}
			w.Flush()
		}
	})

	app.Command("issues", "Displays open Issues for the given repo(s)", func(cmd *cli.Cmd) {
		stateArg := cmd.StringOpt("state", "open", "Display Issues based on their state")
		milestoneArg := cmd.StringOpt("m milestone", "", "Milestone to filter on")
		userArg := cmd.StringOpt("u user", "", "Display only Issues from <user>")
		newArg := cmd.BoolOpt("new", false, "Display Issues opened within the past 24 hours")
		cmd.Action = func() {
			w := tabwriter.NewWriter(os.Stdout, 4, 2, 2, ' ', 0)
			fmt.Fprintf(w, "%s\t%s\t%s\t%s\t%s\n", nc("ISSUE"), nc("LAST UPDATED"), nc("CONTRIBUTOR"), nc("MILESTONE"), nc("TITLE"))
			filter := Filter{
				State: *(stateArg),
				New:   *(newArg),
			}
			if milestoneArg != nil {
				filter.Milestone = *(milestoneArg)
			}
			if userArg != nil {
				filter.User = *(userArg)
			}
			for _, repo := range repoList {
				issues, _ := g.GetIssues(repo, filter)
				g.DisplayIssues(w, repo, issues)
			}
			w.Flush()
		}
	})

	err := app.Run(os.Args)
	if err != nil {
		log.Fatalf(err.Error())
	}
}


utils.go
package main

import (
	"fmt"
	"time"

	"github.com/aybabtme/rgbterm"
)

// shortcut for "no" colour
func nc(s string) string {
	return rgbterm.String(s, 255, 255, 255, 0, 0, 0)
}

// HumanDuration returns a human-readable approximation of a duration
// This function is taken from the Docker project, and slightly modified
// to cap units at days.
// (eg. "About a minute", "4 hours ago", etc.)
// (c) 2013 Docker, inc. and the Docker authors (http://docker.io)
func HumanDuration(d time.Duration) string {
	if seconds := int(d.Seconds()); seconds < 1 {
		return "Less than a second"
	} else if seconds < 60 {
		return fmt.Sprintf("%d seconds", seconds)
	} else if minutes := int(d.Minutes()); minutes == 1 {
		return "About a minute"
	} else if minutes < 60 {
		return fmt.Sprintf("%d minutes", minutes)
	} else if hours := int(d.Hours()); hours == 1 {
		return "About an hour"
	} else if hours < 48 {
		return fmt.Sprintf("%d hours", hours)
	}
	return fmt.Sprintf("%d days", int(d.Hours()/24))
}


github.go
package main

import (
	"fmt"
	"io"
	"os"
	"time"

	"github.com/aybabtme/rgbterm"
	"github.com/google/go-github/github"
	"golang.org/x/oauth2"
)

// GitHubClient -
type GitHubClient struct {
	client *github.Client
}

// Repo -
type Repo struct {
	Org  string
	Name string
}

// Filter -
type Filter struct {
	Milestone string
	User      string
	State     string
	New       bool
}

func (r Repo) String() string {
	return r.Org + "/" + r.Name
}

// NewGitHubClient -
func NewGitHubClient() *GitHubClient {
	ts := oauth2.StaticTokenSource(
		&oauth2.Token{AccessToken: os.Getenv("GITHUB_API_TOKEN")},
	)
	tc := oauth2.NewClient(oauth2.NoContext, ts)

	client := github.NewClient(tc)

	return &GitHubClient{client: client}
}

// FindTeamID -
func (g *GitHubClient) FindTeamID(org string, teamName string) (ID int, err error) {
	var last bool
	var next int
	var teams []*github.Team

	for true {
		t, resp, err := g.client.Organizations.ListTeams(org, &github.ListOptions{Page: next})
		if err != nil {
			return -1, err
		}
		teams = append(teams, t...)
		next = resp.NextPage

		// the _next_ iteration is going to be the last one...
		if last {
			break
		}
		last = resp.NextPage == resp.LastPage
	}

	for _, team := range teams {
		if *(team.Name) == teamName || *(team.Slug) == teamName {
			return *(team.ID), nil
		}
	}
	return -1, fmt.Errorf("Team '%s' not found in org '%s'", teamName, org)
}

// GetRepos -
func (g *GitHubClient) GetRepos(org string, teamName string) (repoList []Repo, err error) {
	teamID, err := g.FindTeamID(org, teamName)
	if err != nil {
		return nil, err
	}

	repos, _, err := g.client.Organizations.ListTeamRepos(teamID, &github.ListOptions{PerPage: 200})
	if err != nil {
		return nil, err
	}

	for _, repo := range repos {
		if *(repo.Owner.Login) == org {
			repoList = append(repoList, Repo{*(repo.Owner.Login), *(repo.Name)})
		}
	}
	return repoList, err
}

// GetPRs -
func (g *GitHubClient) GetPRs(repo Repo, filter Filter) (pulls []*github.PullRequest, err error) {
	pulls, _, err = g.client.PullRequests.List(repo.Org, repo.Name, &github.PullRequestListOptions{
		State:       filter.State,
		ListOptions: github.ListOptions{PerPage: 200},
	})
	var filtered []*github.PullRequest
	if filter.Milestone != "" {
		for _, pull := range pulls {
			if pull.Milestone != nil && *(pull.Milestone.Title) == filter.Milestone {
				filtered = append(filtered, pull)
			}
		}
	} else {
		filtered = pulls
	}

	if filter.User != "" {
		f := filtered
		filtered = []*github.PullRequest{}
		for _, pull := range f {
			if pull.User != nil && *(pull.User.Login) == filter.User {
				filtered = append(filtered, pull)
			}
		}
	}

	if filter.New {
		f := filtered
		filtered = []*github.PullRequest{}
		for _, pull := range f {
			if time.Since(*(pull.CreatedAt)) <= 24*time.Hour {
				filtered = append(filtered, pull)
			}
		}
	}

	return filtered, err
}

// GetIssues -
func (g *GitHubClient) GetIssues(repo Repo, filter Filter) (issues []*github.Issue, err error) {
	issues, _, err = g.client.Issues.ListByRepo(repo.Org, repo.Name, &github.IssueListByRepoOptions{
		State:       filter.State,
		ListOptions: github.ListOptions{PerPage: 200},
	})
	var filtered []*github.Issue
	if filter.Milestone != "" {
		for _, issue := range issues {
			if issue.Milestone != nil && *(issue.Milestone.Title) == filter.Milestone {
				filtered = append(filtered, issue)
			}
		}
	} else {
		filtered = issues
	}

	if filter.User != "" {
		f := filtered
		filtered = []*github.Issue{}
		for _, issue := range f {
			if issue.User != nil && *(issue.User.Login) == filter.User {
				filtered = append(filtered, issue)
			}
		}
	}

	if filter.New {
		f := filtered
		filtered = []*github.Issue{}
		for _, issue := range f {
			if time.Since(*(issue.CreatedAt)) <= 24*time.Hour {
				filtered = append(filtered, issue)
			}
		}
	}

	return filtered, err
}

// DisplayPullRequests -
func (g *GitHubClient) DisplayPullRequests(w io.Writer, repo Repo, pulls []*github.PullRequest) {
	for _, p := range pulls {
		var number = *(p.Number)

		var updated = HumanDuration(time.Since(*(p.UpdatedAt)))
		var contributor = *(p.User.Login)

		var milestone string
		if p.Milestone != nil {
			if time.Since(*(p.Milestone.DueOn)).Seconds() > 0 {
				milestone = rgbterm.String(*(p.Milestone.Title), 255, 0, 0, 0, 0, 0)
			} else {
				milestone = rgbterm.String(*(p.Milestone.Title), 192, 192, 192, 0, 0, 0)
			}
		} else {
			milestone = rgbterm.String("", 192, 192, 192, 0, 0, 0)
		}

		var title = *(p.Title)

		pr := fmt.Sprintf("%s/%s#%d", repo.Org, repo.Name, number)
		fmt.Fprintf(w, "%s\t%s\t%s\t%s\t%s\n", nc(pr), nc(updated), nc(contributor), milestone, nc(title))
	}
}

// DisplayIssues -
func (g *GitHubClient) DisplayIssues(w io.Writer, repo Repo, issues []*github.Issue) {
	for _, issue := range issues {
		var number = *(issue.Number)

		var updated = HumanDuration(time.Since(*(issue.UpdatedAt)))
		var contributor = *(issue.User.Login)

		var milestone string
		if issue.Milestone != nil {
			if time.Since(*(issue.Milestone.DueOn)).Seconds() > 0 {
				milestone = rgbterm.String(*(issue.Milestone.Title), 255, 0, 0, 0, 0, 0)
			} else {
				milestone = rgbterm.String(*(issue.Milestone.Title), 192, 192, 192, 0, 0, 0)
			}
		} else {
			milestone = rgbterm.String("", 192, 192, 192, 0, 0, 0)
		}

		var title = *(issue.Title)

		i := fmt.Sprintf("%s/%s#%d", repo.Org, repo.Name, number)
		fmt.Fprintf(w, "%s\t%s\t%s\t%s\t%s\n", nc(i), nc(updated), nc(contributor), milestone, nc(title))
	}
}
